refactor(kubectl): add ProfileApplier interface for kubectl-debug

knight42 opened the pull request:
Signed-off-by: Jian Zeng <zengjian.zj@bytedance.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref https://github.com/kubernetes/kubectl/issues/1108

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#debugging-profiles
```

knight42 commented:
/cc @verb 

Hi @verb! PTAL

knight42 commented:
@verb a gentle ping. I am looking forward to your comments, how does the implementation looks to you?

verb commented:
@knight42 Thanks for the PR and sorry for the radio silence. This PR is in my queue for this week.

verb reviewed (commented):
We're going to support something like 7 debug profiles, and I'm worried that putting all of that logic in one place is going to grow unwieldy. A profile is pretty basic though. It's only job is to generate a `v1.Pod` that we either create or patch.

@knight42 What do you think about creating a `profiles` package with an interface something like:

```
type Debugger interface {
  NewPod(target runtime.Object) *v1.Pod
  NewEphemeralContainer(target runtime.Object, forPod *v1.Pod) *v1.Pod
}
```

Each profile (i.e. `profiles.General`, `profiles.Legacy`, etc) could then implement the `Debugger` logic and tests separately while the create/patch/update logic remains in debug.go.

verb commented:
/assign

knight42 commented:
@verb Hi, your proposal sounds great! How about we refactor the existing code to adopt this interface in this PR?

verb commented:
> @verb Hi, your proposal sounds great! How about we refactor the existing code to adopt this interface in this PR?

@knight42 sounds good to me, let's do it! :+1: :smiley:

verb commented:
@knight42 double checking: you're working on refactoring this PR now and not waiting on me for anything, correct?

knight42 commented:
@verb Hi！yeah I am working on refactoring the existing code, but now I am on vacation, I'll get back to this on 10-08. I hope I am not blocking your progress.

verb commented:
@knight42 That sounds good to me, enjoy your time off!

verb commented:
> I have added a `Debugger` interface and refactor the existing code to adopt it, but the method signatures are a bit different to the ones you proposed

That's ok, I was only posting pseudocode

> I changed the `NewPod` and `NewEphemeralContainer` to accept a `Option` object as an argument, b/c we change the generated pod or ephermeral container according to some command line args, so I decided to pass them using these Option objects. @verb How does this sound to you?

Makes sense, I think that's fine.

verb commented:
> > I changed the `NewPod` and `NewEphemeralContainer` to accept a `Option` object as an argument, b/c we change the generated pod or ephermeral container according to some command line args, so I decided to pass them using these Option objects. @verb How does this sound to you?
> 
> Makes sense, I think that's fine.

Oh, I didn't notice this was different from `DebugOptions`. Looker deeper now.

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles/options.go:
```diff
@@ -0,0 +1,193 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type DebugPodOptions struct {
```
Can we not just pass `DebugOptions` to the profiles?

knight42 replied:
I think it would cause cyclic import if we make the debugger rely on the `DebugOptions` and I think it might be more elegant to pass the source pod via `DebugPodOptions` than adding an argument to the method `NewPod`.

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles/options.go:
```diff
@@ -0,0 +1,193 @@
+limitations under the License.
+*/
+
+package profiles
```
It's a bit subjective, but packages tend to be singular, and I think it reads a bit better (e.g. `profile.Legacy`, `profile.Unprivileged`)

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles/legacy.go:
```diff
@@ -0,0 +1,63 @@
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type legacyProfile struct {
```
"Profile" is in the package name, so you can omit it here. I would export it (`profile.Legacy`) and define the interface where it's called in `debug.go`.

https://github.com/golang/go/wiki/CodeReviewComments#package-names

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles/options.go:
```diff
@@ -0,0 +1,193 @@
+	SourcePod *corev1.Pod
+}
+
+func (o *DebugPodOptions) Build() (*corev1.Pod, error) {
```
I know it's a draft, but it'd help me review if the https://github.com/golang/go/wiki/CodeReviewComments#doc-comments were already in place.

I think this is the boilerplate for creating a pod that will be common between different profiles. That SGTM and we can keep it even if we get rid of the DebugTypeOptions concept.

knight42 replied:
I am really sorry for confusing you.. I have added some comments, could you take another look?

> I think this is the boilerplate for creating a pod that will be common between different profiles

Exactly!

verb replied:
There's so much boilerplate it makes me think the interface isn't quite right. Thinking about it a bit more, what's interesting about profiles is less about what's common between them and more about the difference between them. Maybe profiles should mutate rather than generate. What do you think about something like https://github.com/verb/kubernetes/commit/9ee09c6e43d10b78eb3dc1b05b461751cee4921e ?

knight42 replied:
@verb Hi! Your alternative looks great, which reduce the code massively. I have updated the code as you said, could you please take another look?

verb reviewed (commented):
@knight42 sorry for the delay, I was traveling. I'll be in the office until code freeze so we can iterate faster.

A few comments, but I think this is the right direction overall.

knight42 committed:
feat(kubectl): add debug profile applier

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

knight42 commented:
@verb Hi! This is just a gentle ping. Now that v1.24 is open, would you mind taking another look?

verb commented:
Hi @knight42, I'm back from the holidays now and will take a look this week. Thanks!

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profile/applier.go:
```diff
@@ -0,0 +1,41 @@
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type Applier interface {
```
It looks like this design progressed from profiles that generate a "debugging object" to one where profiles mutate objects passed to them. I think that makes sense, and it made me realize that all of the `kubectl debug` invocations follow a pattern:

1. Generate a pod to create or update
2. Mutate based on CLI options
3. Mutate based on profile
4. Apply the pod changes

If the addition here is only to mutate via profile, then we can probably use a simpler interface, something like:

```
type applierFunc func(pod *corev1.Pod, containerName string, target runtime.Object) error

type Options struct {
	applier applierFunc
}

// New returns a new Options for the given profile name.
func New(profile string) (*Options, error) {
	switch profile {
	case "legacy":
		return &Options{Legacy}, nil
	}

	return nil, fmt.Errorf("unknown profile: %s", profile)
}

func (o *Options) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
	return o.applier(pod, containerName, target)
}
```

I pushed a full example here: https://github.com/verb/kubernetes/commit/dab06a57016203e645b07086553fe6f3c8af9901

what do you think?

knight42 replied:
@verb Hi! I think it is good and have updated the code.

verb commented:
@knight42 will have time to work on this feature this cycle?

knight42 commented:
> @knight42 will have time to work on this feature this cycle?

@verb Definitely yes! Let's finish it this cycle.

knight42 commented:
/test pull-kubernetes-e2e-kind-ipv6

verb reviewed (changes requested):
@knight42 Thanks for your patience here. I added a few minor suggestions, but overall LGTM. Let's get this merged so we can start iterating on more profiles! :grinning:

verb commented:
> @knight42 Thanks for your patience here. I added a few minor suggestions, but overall LGTM. Let's get this merged so we can start iterating on more profiles! grinning

Oh, and please squash: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#squashing

knight42 commented:
@verb Hi! I believe all your comments has been addressed now, would you mind taking another look?

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profile/options.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+const (
+	LegacyName = "legacy"
```
nit: move to legacy.go and use single line format, `const LegacyName = "legacy"`

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profile/options.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+type applierFunc func(pod *corev1.Pod, containerName string, target runtime.Object) error
+
```
Please add doc comment.

https://github.com/golang/go/wiki/CodeReviewComments#doc-comments

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profile/options.go:
```diff
@@ -0,0 +1,48 @@
+	return nil, fmt.Errorf("unknown profile: %s", profile)
+}
+
+func (o *Options) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
```
Please add doc comment.

https://github.com/golang/go/wiki/CodeReviewComments#doc-comments

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profile/legacy.go:
```diff
@@ -0,0 +1,59 @@
+		pod.Spec.HostPID = true
+		return nil
+	default:
+		return fmt.Errorf("the legacy profile doesn't support objects of type %T", target)
```
nit: `fmt.Errorf("the %s profile ...", LegacyName, ...`

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug_test.go:
```diff
@@ -1255,6 +1271,7 @@ func TestCompleteAndValidate(t *testing.T) {
 				Namespace:      "test",
 				PullPolicy:     corev1.PullPolicy("Always"),
 				ShareProcesses: true,
+				Profile:        "legacy",
```
nit: use `LegacyName` here and below.

verb reviewed (changes requested):
A few more suggestions I didn't notice the first time. Please address, but I think this is ready to ask for @soltysh's review. Thanks!

/lgtm

knight42 commented:
@verb Hi! Sorry for the late response, it seems I somehow missed the notification.. I have addressed your comments now, would you like to take another look?

verb commented:
/retest

verb commented:
/lgtm

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profile/legacy.go:
```diff
@@ -0,0 +1,59 @@
+limitations under the License.
+*/
+
+package profile
```
Nit: I don't see a reason to have separate packge just for profiles, this should be part of just debug package.

knight42 replied:
Updated. cc @verb

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -403,6 +417,10 @@ func (o *DebugOptions) debugByEphemeralContainer(ctx context.Context, pod *corev
 	klog.V(2).Infof("new ephemeral container: %#v", debugContainer)
 	debugPod := pod.DeepCopy()
 	debugPod.Spec.EphemeralContainers = append(debugPod.Spec.EphemeralContainers, *debugContainer)
+	if err := o.profile.Apply(debugPod, debugContainer.Name, pod); err != nil {
```
Shouldn't that move over to `generateDebugContainer`? Alternatively each profile should be applied right after `generate...` invocation. Currently there are inconsistencies.

knight42 replied:
Updated

soltysh reviewed (approved):
Some nits, which would be nice to address. 
/approve
/hold
if you want to get them addressed before the freeze

soltysh commented:
/approve
/hold

soltysh commented:
@knight42 if you manage to get to it by code freeze ping me for re-lgtm, if not feel free to drop the hold tag

verb commented:
@knight42 we won't have time to add additional profiles before code freeze, if we merge this now we'll introduce a new flag that doesn't do anything besides preserve existing behavior. I recommend letting this merge after the freeze when we can add the remaining profiles.

knight42 commented:
> I recommend letting this merge after the freeze when we can add the remaining profiles.

@verb @soltysh I am okay with that 👍

soltysh commented:
> @verb @soltysh I am okay with that +1

Just ping me on slack, when you have that ready, thank you!

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profile_options.go:
```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
Nit: 2022

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
Nit: 2022

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profile_options.go:
```diff
@@ -0,0 +1,48 @@
+// applierFunc is a function that applies a profile to a container in the pod.
+type applierFunc func(pod *corev1.Pod, containerName string, target runtime.Object) error
+
+type ProfileOptions struct {
```
Can we just turn this into an interface?
```golang
type ProfileApplier interface {
    Apply(pod *corev1.Pod, containerName string, target runtime.Object) error
}
```

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profile_options.go:
```diff
@@ -0,0 +1,48 @@
+}
+
+// NewProfile returns a new Options for the given profile name.
+func NewProfile(profile string) (*ProfileOptions, error) {
```
With the above suggestion I'd rename this to `ToProfileApplier` or `NewProfileApplier`

soltysh replied:
With the above suggestion I'd rename this to `ToProfileApplier` or `NewProfileApplier`

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profile_options.go:
```diff
@@ -0,0 +1,48 @@
+/*
```
With the other renames, `profile_applier.go` would also make sense.

soltysh reviewed (commented):
A couple of minor nits and ship it :)

knight42 commented:
/test pull-kubernetes-integration

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -0,0 +1,59 @@
+)
+
+// profileLegacy represents the legacy debugging profile which is backwards-compatible with 1.23 behavior.
+func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object) error {
```
I was more imagining this be a regular type, iow. 
```golang
type LegacyProfile struct {}

func (*LegacyProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
...
```

soltysh reviewed (approved):
/lgtm
/approve

jberkus commented:
This is a feature, shouldn't it have a release note?

@csantanapr

knight42 commented:
> introduce a new flag that doesn't do anything besides preserve existing behavior.

@jberkus This PR introduces a new flag that only preserve existing behavior. The PR that adds more profiles is on the way, I think we could add a release note in that PR.
